src/Tools/Pagination/Paginator.php

```diff
@@ -171,13 +219,30 @@ private function cloneQuery(Query $query): Query
     /**
      * Determines whether to use an output walker for the query.
      */
-    private function useOutputWalker(Query $query): bool
+    private function useOutputWalker(Query $query, bool $forCountQuery = false): bool
     {
-        if ($this->useOutputWalkers === null) {
-            return (bool) $query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) === false;
+        if ($this->useOutputWalkers !== null) {
+            return $this->useOutputWalkers;
         }
 
-        return $this->useOutputWalkers;
+        // When a custom output walker already present, then do not use the Paginator's.
+        if ($query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) !== false) {
+            return false;
+        }
+
+        // When not joining onto *ToMany relations, then do not use the more complex CountOutputWalker.
+        // phpcs:ignore SlevomatCodingStandard.ControlStructures.UselessIfConditionWithReturn.UselessIfCondition
+        if (
+            $forCountQuery
+            && $this->queryStyleAutoDetectionEnabled
+            && ! $this->fetchJoinCollection === false
```

Comment:
Don't compare boolean variable with boolean constants
```suggestion
            && $this->fetchJoinCollection
```

Reply:
It's surprising to me that I left that in the code, despite reviewing it twice myself. Thanks.